### ISSUE TITLE
Loss with multiple GPUs

### ISSUE THREAD

**Issue by soloice** [XMUNLP/XMUNMT#8]
In
https://github.com/XMUNLP/XMUNMT/blob/84d2d70b348be3ea82d13b9b7bb92b239f792a4c/xmunmt/bin/trainer.py#L309, it's simply an arithmetic average of losses collected from different devices.
This code is perfect for a single-device setting. But if you have multiple devices and different devices afford different amount of computation in a mini-batch (say, device 0 deals with shorter sentences, while device 1 faces longer ones), the loss calculated in this way will be biased. (Nevertheless, this carelessness should not introduce too much impact, unless the data distribution is extremely unbalanced.)

Ideally, it should be a weighted average of all losses, and the weights are number of valid tokens in each device, shouldn't it?

**Comment by Playinf** [comment]
The current version is equivalent to averaging gradients computed by each device, which is a common practice for multi-device setting. As you pointed out, the behavior is slightly different from a single device setting. However, computing weights will incur additional performance cost and the data processing procedure ensures that the sentences in a batch will have similar length.
 The optimizations of neural networks are complex and people often favor a simple and efficient implementation. For example, the word2vec implementation uses asynchronous update even though that mechanism is different from mathematical inductions.

**Comment by soloice** [comment]
Yep, and in my experiment `word2vec` with a single thread outperforms its multi-threading counterpart, lol~

By the way, how to extend this framework to more NMT architectures? I think the canonical way is to write a class for each different model architecture and put it under `models` submodule, then run `bin/train.py` and `bin/translator.py` with command line parameter `--model <model-name>`.

But what if the new models has a different data feeding mechanism? For example, if I want to combine NMT with GAN (generative adversarial network), and want to update G (the generator in GAN, typically an NMT model) and D (the discriminator in GAN, maybe a CNN) in an interleaved manner, what's the best practice to achieve this? Am I supposed to write a new `train.py` for my model? Or merge all train_ops/data pipeline into a single one ? Or set up another repo for this purpose?
Thanks a lot.

**Comment by Playinf** [comment]
I think the easiest way to write a GAN model is to use the `tf.contrib.gan.estimator.GANEstimator` framework. Unfortunately, I don't have many experiences on GANs. The interface of `tf.contrib.gan.estimator.GANEstimator` assumes one input function, so I think the approach to merge all train_ops/data pipeline should be appropriate.

**Comment by soloice** [comment]
Thanks for your advice. I'll have a try.
